### PR TITLE
Mark GEP as inbounds (memref) and unlock derivative of indexing/slicing

### PR DESCRIPTION
**Pull request by rmoyard**
**Description of the Change:**

We mark GEPOp inbounds when they do not have a null ptr as defining operation, this allows Enzyme to do a better activity analysis and it unlocks the derivative of indexing arrays.

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-57426]

**Comment by rmoyard** [comment]
[sc-55109]

File: mlir/test/Catalyst/GEPInbounds.mlir [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
```
Comment (dime10) [inline mlir/test/Catalyst/GEPInbounds.mlir:1]:
```suggestion
// Copyright 2024 Xanadu Quantum Technologies Inc.
```

File: mlir/include/Catalyst/Transforms/Passes.td [hunk 2]
```diff
@@ -124,4 +124,14 @@ def AddExceptionHandlingPass : Pass<"add-exception-handling"> {
     ];
 }
 
+def GEPInboundsPass : Pass<"gep-inbounds"> {
+    let summary = "Mark GEPOp inbounds.";
+
+    let dependentDialects = [
+        "mlir::LLVM::LLVMDialect",
+    ];
+
```
Comment (dime10) [inline mlir/include/Catalyst/Transforms/Passes.td:133]: We don't create any llvm ops, so the dialect is not technically a dependent dialect :)
```suggestion
```

**Review by dime10 — APPROVED** [review]
Great 🥳

**Pull request merged**